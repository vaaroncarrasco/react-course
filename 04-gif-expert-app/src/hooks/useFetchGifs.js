// ? Hooks are just functions -> can use state/effect/reducer/context etc
// Functional components most of the time use -> react import is just for JSX
// * hooks start with useCustomHook

import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs";

export const useFetchGifs = ( category ) => {
    const [state, setState] = useState({
        data: [],
        loading: true
    });

    // useEffect -> only run code block when component is rendered for the first time
    useEffect(() => { // * useEffect can not be async -> use PROMISE .then() instead
        getGifs( category )
            .then(imgs => {
                setState({
                    data: imgs,
                    loading: false
                });
            });
    }, [category]); // ? run when dependency change -> if none [] run code-block only once

    return state; // {  data: [10 imgs], loading: false };

}