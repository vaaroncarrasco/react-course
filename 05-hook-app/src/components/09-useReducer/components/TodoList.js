import React from 'react'
import { TodoListItem } from './TodoListItem'
import PropTypes from 'prop-types';

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
    return (
        <ul className="list-group list-group-flush">
            {
                todos.map( (todo, index) => ( // * () => () -> () implicitly returning an object
                    <TodoListItem
                        key={ todo.id }
                        todo={ todo }
                        index={ index }
                        handleDelete={ handleDelete }
                        handleToggle={ handleToggle }
                    />
                ))
            }
        </ul>
    )
}

TodoList.propTypes = {
    todos: PropTypes.array.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}
