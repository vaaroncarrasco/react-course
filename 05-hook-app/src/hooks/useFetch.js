import { useEffect, useRef, useState } from 'react';

export const useFetch = ( url ) => {

    // * set reference on something to check if its mounted
    const isMounted = useRef(true);

    const [state, setState] = useState({ data: null, loading: true, error: null });

    // * check if mounted to execute functionality
    useEffect( () => {

        return () => { isMounted.current = false; }

    }, []);

    // ? Can not set state on un-mounted element
    useEffect( () => {

        setState({ data: null, loading: true, error: null });

        fetch( url )
            .then( resp => resp.json() )
            .then( data => {
                if ( isMounted.current ) {
                    setState({
                        loading: false,
                        error: null,
                        data
                    });
                }
            })
            .catch( () => {
                setState({
                    data: null,
                    loading: false,
                    error: 'No se pudo cargar la data :(',
                });
            });

    }, [url]);

    return state;

}
