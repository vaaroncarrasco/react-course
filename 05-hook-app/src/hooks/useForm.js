import { useState } from "react"

export const useForm = (initialState = {}) => {

    const [values, setValues] = useState(initialState);

    const reset = () => {
        setValues ( initialState );
    }

    const handleInputChange = ({ target }) => {
        setValues({
            ...values,
            [ target.name ]: target.value
        });
    }

    return [ values, handleInputChange, reset ];

}

// Each <input onChange={ handleInputChange }/> calls for its input value name/email -> calling twice
// setValues({
    // destructure state object ...{}, spread all props in case a prop aint changing
    // setting key name -> objPropName: value
    // name="name" value="formState.name" // name="email" value="formState.email"
//     ...values,
//     [ target.name ]: target.value // name: 'value'
// });