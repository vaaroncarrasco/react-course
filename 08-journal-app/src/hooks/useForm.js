import { useState } from 'react';


export const useForm = ( initialState = {} ) => {

    const [values, setValues] = useState(initialState);

    // to use new state
    const reset = ( newFormState = initialState ) => {
        setValues( initialState );
    }


    const handleInputChange = ({ target }) => {

        setValues({
            ...values,
            [ target.name ]: target.value
        });

    }

    return [ values, handleInputChange, reset ];

}