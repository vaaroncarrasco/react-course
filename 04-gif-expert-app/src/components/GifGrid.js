import React from 'react'
import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {

    // Every time theres a change in component, it runs the whole code all over again and again -> useEffect fixes this by calling functions once

    const { data:images, loading } = useFetchGifs( category );

    return (
        <>
            <h3> { category } </h3>

            { loading && <p>Loading...</p> }

            <div className="card-grid">
                {
                    images.map(img => (
                        <GifGridItem
                            key={ img.id }
                            // * spreading img{} props and sending em all at once instead of prop={ prop }
                            { ...img }
                            // img={ img }
                        />
                    ))
                }
            </div>
        </>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired
}