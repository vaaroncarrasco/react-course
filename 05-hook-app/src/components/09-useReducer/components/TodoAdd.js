import React from 'react'
import { useForm } from '../../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {

    const [ { description }, handleInputChange, reset ] = useForm({ // useForm returns [values{}, handleInputChange]
        // same as name="description" in html attributes
        description: '',
    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if ( description.trim().length <= 1) return;

        // * new object
        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false
        }

        // * dispatch sends action to reducer
        handleAddTodo( newTodo );
        reset();

    }

    return (
        <>
            <h4>Agregar ToDo</h4>
            <hr />

            <form onSubmit={ handleSubmit }>
                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Aprender..."
                    autoComplete="off"
                    value={ description }
                    onChange={ handleInputChange }
                />

                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                >
                    Agregar
                </button>

            </form>
        </>
    )
}
