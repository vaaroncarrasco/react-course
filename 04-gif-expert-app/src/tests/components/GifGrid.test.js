import React from 'react';
import '@testing-library/jest-dom'

import { shallow } from 'enzyme';
import { GifGrid } from '../../components/GifGrid';

// ? simulating / mocking fetched data
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs')

describe('testing GifGrid', () => {

    const category = 'Game of Thrones';

    test('should match snapshot with empty mocked data', () => {

        // * mock the returned value
        useFetchGifs.mockReturnValue({
            data: [],
            loading: true
        });

        const wrapper = shallow( <GifGrid category={ category }/>) ;
        expect( wrapper ).toMatchSnapshot();
    });

    test('should show gifs loaded from useFetchGifs', () => {

        const gifs = [{
            id: 'ABC',
            url: 'https://localhost/cualquier/cosa.jpg',
            title: 'test title'
        }];

        // * mock the returned value
        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false
        });

        const wrapper = shallow( <GifGrid category={ category }/> );

        expect( wrapper ).toMatchSnapshot();
        expect( wrapper.find('p').exists() ).toBe( false );
        expect( wrapper.find('GifGridItem').length ).toBe( gifs.length ); // find <GifGridItem /> component in snapshot

    });

});