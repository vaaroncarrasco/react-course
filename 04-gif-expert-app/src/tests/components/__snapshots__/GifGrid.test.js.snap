// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing GifGrid should match snapshot with empty mocked data 1`] = `
<Fragment>
  <h3>
     
    Game of Thrones
     
  </h3>
  <p>
    Loading...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;

exports[`testing GifGrid should show gifs loaded from useFetchGifs 1`] = `
<Fragment>
  <h3>
     
    Game of Thrones
     
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="test title"
      url="https://localhost/cualquier/cosa.jpg"
    />
  </div>
</Fragment>
`;
